File: EcoGo/src/main/java/com/example/EcoGo/dto/BadgeDto.java
package com.example.EcoGo.dto;

import com.example.EcoGo.model.Badge;

import java.util.Map;

public class BadgeDto {

    private String badgeId;
    private Map<String, String> name;
    private Map<String, String> description;
    private Integer purchaseCost;
    private String category;
    private String subCategory;
    private String acquisitionMethod;
    private Double carbonThreshold;
    private BadgeIconDto icon;
    private Boolean isActive;

    public Badge toEntity() {
        Badge badge = new Badge();
        badge.setBadgeId(this.badgeId);
        badge.setName(this.name);
        badge.setDescription(this.description);
        badge.setPurchaseCost(this.purchaseCost);
        badge.setCategory(this.category);
        badge.setSubCategory(this.subCategory);
        badge.setAcquisitionMethod(this.acquisitionMethod);
        badge.setCarbonThreshold(this.carbonThreshold);
        if (this.icon != null) {
            Badge.BadgeIcon badgeIcon = new Badge.BadgeIcon();
            badgeIcon.setUrl(this.icon.getUrl());
            badgeIcon.setColorScheme(this.icon.getColorScheme());
            badge.setIcon(badgeIcon);
        }
        badge.setIsActive(this.isActive);
        return badge;
    }

    public String getBadgeId() {
        return badgeId;
    }

    public void setBadgeId(String badgeId) {
        this.badgeId = badgeId;
    }

    public Map<String, String> getName() {
        return name;
    }

    public void setName(Map<String, String> name) {
        this.name = name;
    }

    public Map<String, String> getDescription() {
        return description;
    }

    public void setDescription(Map<String, String> description) {
        this.description = description;
    }

    public Integer getPurchaseCost() {
        return purchaseCost;
    }

    public void setPurchaseCost(Integer purchaseCost) {
        this.purchaseCost = purchaseCost;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getSubCategory() {
        return subCategory;
    }

    public void setSubCategory(String subCategory) {
        this.subCategory = subCategory;
    }

    public String getAcquisitionMethod() {
        return acquisitionMethod;
    }

    public void setAcquisitionMethod(String acquisitionMethod) {
        this.acquisitionMethod = acquisitionMethod;
    }

    public Double getCarbonThreshold() {
        return carbonThreshold;
    }

    public void setCarbonThreshold(Double carbonThreshold) {
        this.carbonThreshold = carbonThreshold;
    }

    public BadgeIconDto getIcon() {
        return icon;
    }

    public void setIcon(BadgeIconDto icon) {
        this.icon = icon;
    }

    public Boolean getIsActive() {
        return isActive;
    }

    public void setIsActive(Boolean isActive) {
        this.isActive = isActive;
    }

    public static class BadgeIconDto {
        private String url;
        private String colorScheme;

        public String getUrl() {
            return url;
        }

        public void setUrl(String url) {
            this.url = url;
        }

        public String getColorScheme() {
            return colorScheme;
        }

        public void setColorScheme(String colorScheme) {
            this.colorScheme = colorScheme;
        }
    }
}

File: android-app/app/src/main/kotlin/com/ecogo/data/dto/BadgeDto.kt
package com.ecogo.data.dto

import com.google.gson.annotations.SerializedName
import java.util.Date

/**
 * Badge - 统一的徽章/服饰数据模型
 * 后端使用同一个表存储 badge 和 cloth
 */
data class BadgeDto(
    @SerializedName("id")
    val id: String? = null,

    @SerializedName("badgeId")
    val badgeId: String,

    @SerializedName("name")
    val name: Map<String, String>?,

    @SerializedName("description")
    val description: Map<String, String>?,

    @SerializedName("purchaseCost")
    val purchaseCost: Int?,

    @SerializedName("category")
    val category: String?,  // "badge" 或 "cloth"

    @SerializedName("subCategory")
    val subCategory: String?,  // cloth: head/face/body, badge: rank

    @SerializedName("acquisitionMethod")
    val acquisitionMethod: String?,

    @SerializedName("carbonThreshold")
    val carbonThreshold: Double?,

    @SerializedName("icon")
    val icon: BadgeIcon?,

    @SerializedName("isActive")
    val isActive: Boolean?,

    @SerializedName("createdAt")
    val createdAt: Date? = null
) {
    data class BadgeIcon(
        @SerializedName("url")
        val url: String?,

        @SerializedName("colorScheme")
        val colorScheme: String?
    )

    fun isCloth(): Boolean = category == "cloth"
    fun isBadge(): Boolean = category == "badge"
    fun getEnglishName(): String = name?.get("en") ?: "Unknown"
}

/**
 * UserBadge - 用户持有的徽章/服饰
 */
data class UserBadgeDto(
    @SerializedName("id")
    val id: String? = null,

    @SerializedName("userId")
    val userId: String,

    @SerializedName("badgeId")
    val badgeId: String,

    @SerializedName("unlockedAt")
    val unlockedAt: Date?,

    @SerializedName("isDisplay")
    val isDisplay: Boolean,

    @SerializedName("createdAt")
    val createdAt: Date? = null
)

/**
 * 购买请求
 */
data class PurchaseRequest(
    @SerializedName("user_id")
    val userId: String
)

/**
 * 切换佩戴请求
 */
data class ToggleDisplayRequest(
    @SerializedName("user_id")
    val userId: String,

    @SerializedName("is_display")
    val isDisplay: Boolean
)

/**
 * 用户装备信息
 */
data class UserOutfitDto(
    val head: String? = null,
    val face: String? = null,
    val body: String? = null,
    val badge: String? = null
)